src/whoami_assets/public/WhoamiProcess.ts

```diff
@@ -111,43 +140,82 @@ async function authenticate(this: Pick<typeof globalThis, "document">) {
   }
   // We either have a new candidate session, or one from storage.
   // Either way, we want to use it.
-  const useSessionCommand = AuthenticatorSession(secretSession);
-  authenticator.useSession(useSessionCommand);
+  const publicSession = AuthenticatorSession(secretSession);
+  authenticator.useSession(publicSession);
+
+  const icAgentIdentity = createSessionIdentity(secretSession)
+  options.onIdentity(icAgentIdentity)
 }
 
 /**
- * Listen for changes to the authenticated Identity.
- * For each identity:
- * * log the new identity principalHex
- * * call testWhoamiContract to test the new identity
- * @param this Window
- * @param options options
- * @param options.events - where to dispatch/listenFor DOM events
- * @param options.render - call to render something where the end-user can see it.
+ * Create an IC HttpAgent to act on behalf of the session
+ * @param identity - identity to use to sign icp requests 
  */
-async function handleEachIdentity(
-  this: Pick<typeof globalThis, "document">,
-  options: {
-    events: Pick<EventTarget, "addEventListener" | "dispatchEvent">;
-    render: Render;
-  }
-) {
-  const log = makeLog("handleEachIdentity");
-  const { events, render } = options;
-  const identityGenerator = IdentitiesIterable(events);
-  for await (const identity of identityGenerator) {
-    const principalHex =
-      identity.type === "AnonymousIdentity"
-        ? new AnonymousIdentity().getPrincipal().toHex()
-        : Principal.selfAuthenticating(blobFromHex(identity.publicKey)).toHex();
-    log("debug", "identity", {
-      ...identity,
-      principalHex,
-    });
-    await testWhoamiContract.call(this, { render });
+function createSessionAgent(identity: PublicIdentity) {
+  console.debug('createSessionAgent', { identity })
+  const agent = new HttpAgent({
+    host: '',
+    identity,
+  });
+  return agent;
+}
+
+/**
+ * Given a session, return a SignIdentity
+ * @param secretSession - session
+ * @returns SignIdentity corresponding to info in session
+ */
+function createSessionIdentity(secretSession: Readonly<Session>): PublicIdentity {
+  console.debug('createSessionIdentity', { secretSession })
+  const { authenticationResponse } = secretSession;
+  if ( ! authenticationResponse) {
+    return new AnonymousIdentity();
   }
+  const hash = new URL(authenticationResponse).hash?.slice(1);
+  console.log('hash', hash)
+  const icidResponse = response.fromQueryString(new URLSearchParams(hash))
+  const parsedBearerToken = response.parseBearerToken(icidResponse.accessToken);
+  const delegationIdentity = DelegationIdentity.fromDelegation(
+    SessionSignIdentity(secretSession),
+    DelegationChain.fromJSON(JSON.stringify(parsedBearerToken)),
+  );
+  return delegationIdentity;
```

Comment:
This is so common we should probably export a better function that goes from authenticationResponse URL to DelegationChain